gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1503,19 +1503,33 @@ private Credential getCredential(Configuration config)
    */
   private static Optional<Credential> getImpersonatedCredential(
       Configuration config, Credential credential) throws IOException {
+    GoogleCloudStorageOptions options =
+        GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
+            .build()
+            .getCloudStorageOptions();
+
     String serviceAccountToImpersonate =
         IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
             .withPrefixes(CONFIG_KEY_PREFIXES)
             .get(config, config::get);
+    if (isNullOrEmpty(serviceAccountToImpersonate)) {
+      String serviceAccountFromUserOrGroup =
+          getMatchedServiceAccountToImpersonateFromGroup(
+                  options.getUserImpersonationServiceAccounts(),
+                  ImmutableList.of(UserGroupInformation.getCurrentUser().getShortUserName()))
+              .orElse(
+                  getMatchedServiceAccountToImpersonateFromGroup(
+                          options.getGroupImpersonationServiceAccounts(),
+                          ImmutableList.copyOf(
+                              UserGroupInformation.getCurrentUser().getGroupNames()))
+                      .orElse(null));
```

Comment:
Can we reformat so the indentations for user-match --> group-match --> null are matched.

Reply:
Agreed it's not readable enough. Extracted those two unreadable method calls to two variables instead so logic will be more clear.